### ISSUE TITLE
_flattenBehaviorsList in Polymer micro should return a set of unique behaviors rather than a concatenation of all behaviors

### ISSUE THREAD

**Issue by codegist** [Polymer/polymer#1894]
I came into an issue while using behaviors in my project where I end up having modules that have their in memory behaviors array containing multiple instance of the same behavior.
This is partly due to how I use behaviors, but I also think _flattenBehaviorsList should make sure the final array of behaviors a module is assigned to is a set of unique behaviors.

Here's why I end up with duplicate behaviors:
Module A has behaviors B1 and B2
Behaviors B1 itself has behavior B2.

Module A in memory behaviors array currently ends up being [B1, B2, B2].

Now I did write my code that way because from Module A POV, I do not think it should make the assumption that behavior B1 already it "behaviored" with B2. If Module A needs behavior B2 functionality, it just makes it clear in its behaviors list rather than relying on B1 already having B2.

This is a problem because on any property change with observers on behavior B2, change callback will be called twice! Which in my implementation case ends up being two HTTP request being fired.

**Comment by codegist** [comment]
It is a fairly easy fix, here's how I patched it in the meantime. (added _set function and wrapped _prepBehaviors's_flattenBehaviorsList call into _set)

```javascript
_set:function(array){
    var setArr = [];
    for (var i = array.length-1; i > 0; i--) {
        if(setArr.indexOf(array[i]) == -1) {
            setArr.push(array[i]);
        }
    };
    return setArr;
},
_prepBehaviors: function () {
    if (this.behaviors.length) {
        this.behaviors = this._set(this._flattenBehaviorsList(this.behaviors));
    }
    this._prepAllBehaviors(this.behaviors);
},
```

**Comment by codegist** [comment]
Created a pull request for it: #1910

**Comment by cdata** [comment]
> I also think _flattenBehaviorsList should make sure the final array of behaviors a module is assigned to is a set of unique behaviors.

:+1: for this. The taxonomy of the Polymer Elements catalogue is becoming sufficiently interconnected that we are starting to run into this problem from time to time.

**Comment by sjmiles** [comment]
The ordering of behaviors is important, so there is no "correct flattened order" in general. The intent is that users of behaviors need to be careful about how they are applied.

**Comment by codegist** [comment]
Yes I agree for the ordering bit.. Well I think behaviors can be used in a lots of different manners, and probably in ways that weren't thought of initially. Let's say one create a base behavior with helpers methods (for instance helper method for computing classes/styles/conditions since filters and expressions are not supported anymore in 1.0) and say, an ID incremented on "create" callback, assign to every component/behaviors that uses it. Now in a project, the hierarchy/complexity of behavior relationship can and will become fairly complex and in the case of a common/base behavior, that would be used as a base for any behavior/component implemented in said project, the user will end up in the use case I've exposed above, where that base behavior will appears many times in the array of behavior a module in assigned to. And to go on with this example of incremented ID this base behavior exposes, this ID will be incremented as many times as the base behavior is present/duplicated in the module flattened behavior array. Plus all callback/observers attached to that base behavior will be called as many times as well.

I believe this will create bugs in applications and confusion in devs because it is not something easily spotable at first. Looking at the pros and cons, I think flattening and removing duplicates from the behaviors list is better than leaving it as is. As for the ordering, a choice should be made and diligently documented as such. I think if one ends up with a list of behaviors as the following:

[[A,B],[C,A],B,D,A,C]
the list should be transformed to this
[B,D,A,C]
eg: removing any instance of duplicated behaviors from the list and keeping the last instance of it.

Other programming language are also dealing with this kind of usecase, like Scala with Traits where multiple traits have the same method. The method of the last traits take precedence and wins of the other duplicated ones.

**Comment by justinfagnani** [comment]
There can be a well-defined ordering of behaviors with duplicates, along with restrictions on behavior application and ordering that simplify reasoning (and hopefully implementation):

 * A behavior is only be applied once per element, no matter how many times it's included in the nested behaviors arrays.
 * Nested behavior arrays cannot include _conflicting_ orderings between two behaviors. That is, they cannot include both `[A, B]` and `[B, A]`, or `[A, B, A]` or `[[A, B], A]`. This means behaviors must form a DAG and therefore it's possible to perform a topological sort on the behaviors arrays.
 * The _first_ instance of a behavior is the one applied. This defines the topological sort and preserves dependencies on ordering between behaviors defined in the same array (`[A, B]` -> `B` depends on `A`) and nested arrays `[[A, B], C]` -> `C` depends on `A` and `B`.

I believe this is consistent with languages that allow mixin's to inherit (like Scala's traits) or that support multiple inheritance (like Python's method resolution ordering: http://python-history.blogspot.com/2010/06/method-resolution-order.html )

The example that @codegist gives should be an error because of conflicts in `A`/`B`, `A`/`C`, and `B`/`C`. `[C, A, B, [A, B], [C, A], D]` would be a valid reordering which would flatten to `[C, A, B, D]` (though hopefully the element author would know that they get A, B and C from the nested arrays).

**Comment by justinfagnani** [comment]
> Other programming language are also dealing with this kind of usecase, like Scala with Traits where multiple traits have the same method. The method of the last traits take precedence and wins of the other duplicated ones.

@codegist Note that the order of Polymer behavior arrays are reversed compared to mixins in most languages. In Scala the right most trait in a list is the "oldest", and the left most trait has the highest precedence. In Polymer the _left_ most behavior is the "oldest", and the _right_ most behavior take precedence. This is why it's important to choose the first instance of a behavior.

Also note that methods on the prototype take precedence over behaviors. This is the same as regular mixin application as long as elements don't support inheritance. Once elements do support inheritance it'll be important that behaviors applied to the "subclass" take precedence over the "superclass".

**Comment by cdata** [comment]
:+1: @justinfagnani Thanks for providing a concise explanation of how things might be improved.

_TL;DR for below is that de-duplication would not make behaviors perfect, but it would be an improvement over the current approach that allows duplication._

My general sentiment is that the world with de-duplication is not necessarily more predictable than the world without it. For example, please consider the following contrived behavior hierarchy:

```javascript
// Common base behavior that implements two methods:
var A = {
  foo: function() {
    console.log('A.foo');
  },
  bar: function() {
    console.log('A.bar');
  }
};

// A derivative behavior that overrides both methods of A:
var B = {
  foo: function() {
    console.log('B.foo');
  },
  bar: function() {
    console.log('B.bar');
  },
  baz: function() {
    this.foo();
    console.log('B.baz');
  }
};

// Another derivative behavior that overrides only one method of A,
// and also calls the "super" implementation:
var C = {
  foo: function() {
    A.foo.apply(this, arguments);
    console.log('C.foo');
  }
};

// Distributed versions of behaviors B and C:
var BehaviorB = [A, B];
var BehaviorC = [A, C];

// Ultimate application of behaviors B and C. Note that the lineage of behaviors
// B and C is not immediately obvious to the applier of the behavior:
Polymer({
  is: 'applies-b-and-c',

  behaviors: [BehaviorB, BehaviorC]
});
```

Today, we allow duplication of behaviors in the hierarchy, and the order of the above application will come out as `[A, B, A, C]`. The practical outcome of this is that behavior `B`'s overridden methods `foo` and `bar` will be cancelled out by the subsequent re-application of behavior `A`. In order to fix this problem, I have to re-architect the dependencies of my dependencies, which is time consuming at best and impractical at worst.

Alternatively, if we de-duplicate behaviors according to @justinfagnani 's proposal, the final order of behaviors becomes `[A, B, C]`. This arrangement has a different but similarly unpredictable property: `C.foo`'s intent is to call the "super" behavior's implementation. The relevant "super" behavior of `C` has changed from `A` to `B`, but since `C` only knows about `A`, `C` calls `A`'s implementation directly. `B.foo` is effectively skipped from the hierarchy. In some cases this is fine, but in some cases the result is unexpected (e.g., event handlers).

My opinion - fueled as it is by experience - is that the drawback of de-duplication outlined above is much smaller than that of allowing duplication.

**Comment by cdata** [comment]
Alternatively, you could expressly disallow multiple applications of a behavior (by throwing an exception when it occurs, perhaps). But then behaviors could no longer be treated like mixins.

**Comment by codegist** [comment]
While I do agree the discussion of the aspect of method override and/or priority is valid and what have been exposed is sensible, the reason I originally opened this issue was about another aspect. I didn't even noticed that behaviors where duplicated until I ran into the weird behavior where in case of duplicated behaviors, where those duplicated behaviors registered observers on properties changes, those observers where called as many time as they were duplication for the the same property change. 
eg:

```javascript
var A = {
   properties:{
       someProperty:{
              observer:"somePropertyChanged"
       }
   },
   someProperty:function(){}
}
```

When, for any hierarchy complexity reason, a module is assigned behavior A multiple times, eg making A duplicated, any change on "someProperty" will cause "someProperty()" to be called as many time as A is duplicated.

To me this seems to be wrong and confusing, and my first thought was then to ensure that the behavior list was de-duplicated which corrected that observer callback to be called more than once for a single change.
Now after this discussion it might not be the correct solution (de-duplicating behaviors), but I still think that for what regards properties change listeners, when there are duplicated behavior, callback should be called just once.

**Comment by justinfagnani** [comment]
@cdata I don't think your example is a problem at all. Since ES5 doesn't have `super` you either have to delegate to another function explicitly, like you do in `A.foo.apply(this, arguments)`, or follow one of the "super" patterns where super is passed in to some closure environment.

In the explicit delegation approach, I'd say this is par for the course in ES5. Depending on the case it either works fine, or if you're trying to emulate `super` and "superclass" is not what you expect something could behave oddly. Note that the worst that happens with my ordering proposal is that you skip over some intervening behaviors. Given that Polymer takes care of calling all the lifecycle methods in each behavior and doesn't rely on a `super`-like thing, I don't think this is an issue in practice.

It's important to emphasize that with behaviors you really do not know your "superclass" at definition time - it could be any element applying the behavior, with any number of other behaviors applied first. Explicit delegation should only be used when you know that you're not really trying to emulate `super` in a language that doesn't have it. Behaviors should be able to rely on one another, and even their partial ordering (before and after relationships), but not on their exact ordering (parent and child relationships).

Now, I do think it'd be great to allow behaviors to delegate directly to their preceding behavior, or finally the element, as a general principle. It's fairly easy to do if behaviors are functions that accept their predecessor:

```language-javascript
// Another derivative behavior that overrides only one method of A,
// and also calls the "super" implementation:
var C = function(_super) {
  return {
    foo: function() {
      _super.foo.apply(this, arguments);
      console.log('C.foo');
    }
  };
};
```

In this approach each behavior is really a factory. Polymer would be responsible for invoking the factory with the result of applying the previous behavior, or initially the prototype just above the custom element.

**Comment by devinivy** [comment]
I agree with @justinfagnani– a topological sort will do the trick here to ensure nested behaviors have a sensible order when culled into a flattened, deduplicated list.  Would be happy to help.

**Comment by TimvdLippe** [comment]
This has been fixed in Polymer 2 for legacy behaviors (https://github.com/Polymer/polymer/blob/374d4079e215a8246e3981e2f0c58edc44a1239e/lib/legacy/class.html#L109-L133) and also using mixins (https://github.com/Polymer/polymer/blob/374d4079e215a8246e3981e2f0c58edc44a1239e/lib/utils/mixin.html#L33-L69)
